hybrid/resource/wind_resource.py

```diff
@@ -163,5 +183,30 @@ def data(self, data_file):
         """
         Sets the wind resource data to a dictionary in SAM Wind format (see Pysam.ResourceTools.SRW_to_wind_data)
         """
-
-        self._data = SRW_to_wind_data(data_file)
+        if self.api.lower() == 'nrel':
+            self._data = SRW_to_wind_data(data_file)
+        elif self.api.lower() == 'nasa':
+            # This methodology assumes the NASA POWER API is called with all arguments (wind-surface, wind-elevation, site-elevation) as detailed in example below
+            # https://power.larc.nasa.gov/api/temporal/hourly/point?start=20120101&end=20121231&latitude=35.2018&longitude=-101.9450&community=RE&parameters=T2M&format=srw&wind-surface=vegtype_4&wind-elevation=80&site-elevation=80
+ 
+            data_dict = SRW_to_wind_data(data_file)
+            self._data = {}
+            full_data = np.array(data_dict['data'])
+            num_datapoints = full_data.shape[0]
+            self._data['data'] = np.zeros((num_datapoints, 4))
+            self._data['data'][:, 0] = full_data[:, 0]  # Grab the temp at 2 m
+            self._data['data'][:, 1] = full_data[:, 9]  # Grab the corrected pressure at user specified hub height
+            self._data['data'][:, 2] = full_data[:, 8]  # Grab the corrected speed at user specified hub height
+            self._data['data'][:, 3] = full_data[:, 7]  # Grab the direction at 50 m
+            self._data['heights'] = [self.hub_height_meters] * 4
+            self._data['fields'] = [1, 2, 3, 4]
+            if self.hub_height_meters == 10:
+                # overwrites 50 m direction data with 10 m direction data
+                self._data['data'][:, 3] = full_data[:, 5]      
+            # else:
+            #     height_diffs = np.abs(np.array(data_dict['heights']) - self.hub_height_meters)
+            #     indices = np.where(height_diffs == height_diffs.min())
+            
+            self._data['data'] = self._data['data'].tolist()
```

Comment:
@dguittet yea let's talk about it. So this post processing is only necessary if the specified turbine hub height is not within 10m of the 10m or 50m data points returned by NASA. Therefore if we are looking at hubheights <59 meters no post processing of the data dictionary is required and SAM can run with what is returned by ResourceTools.py (the approved PySAM PR version).

Within the scope of HOPP we are looking at hub heights > 60m so this postprocessing is necessary for our HOPP runs. Given others may use this API outside of HOPP we thought the best approach was to return all data available and let the user / application post process for their needs before passing the data to SAM/SSC. Additionally if in the future NASA can provide all data points at any hub height than no post processing would be needed at all.

However there is no foreseeable timeline for all data points at any hub height being provided by NASA and HOPP is currently the only application using NASA POWER data. I'm leaning toward keeping this out of PySAM for now because 1) I'm pretty limited on bandwidth, I'm closing on 2 weeks until my SULI internship report and presentation are due and 2) the post processing inherently makes some assumptions / simplifications (using data outside of the typical 10m range and overwriting all data point 'heights' to the specified hub height) to allow SAM to process the data, I want to avoid people using this data without fully understanding the simplifications that made it possible.

Let me know your thoughts, happy to talk through this on a call if it is easier.

Reply:
Ah I see now. Yes you are right to have this separate from SRW_to_wind_data since SRW_to_wind_data doesn't know anything about the hub height. It's job is just to convert the file into a dictionary.

While I could see it beneficial to have a function in PySAM.ResourceTools that is just a "expand nasa wind data to desired hub height" function, which replicates the post processing you have here, I agree that keeping it in HOPP where it is more tightly controlled makes good sense.

Still, I do propose moving this functionality to a separate function in wind_resource that is called after `SRW_to_wind_data` is called. Perhaps there should be some checks like "if hub height is not within 10 etc" so the function can return early if not necessary. I could see this function being useful if called in a separate workflow.

Reply:
moved over